services/brig/test/unit/Test/Brig/Calling.hs

```diff
@@ -0,0 +1,206 @@
+  -- We don't want to stop the loop before it has written to the sftServers IORef
+  void $ retryEvery10MicrosWhileN 2000 (== NotDiscoveredYet) (readIORef (sftServers sftEnv))
+  Async.cancel discoveryLoop
+
+  actualServers <- readIORef (sftServers sftEnv)
+  assertEqual "servers should be the ones read from DNS" (Discovered returnedEntries) actualServers
+  pure sftEnv
+
+testDiscoveryLoopWhenUnsuccessful :: IO ()
+testDiscoveryLoopWhenUnsuccessful = do
+  fakeDNSEnv <- newFakeDNSEnv (\_ -> SrvNotAvailable)
+  sftEnv <- mkSFTEnv (SFTOptions "foo.example.com" Nothing (Just 0.001))
+
+  discoveryLoop <- Async.async $ runM . ignoreLogs . runFakeDNSLookup fakeDNSEnv $ sftDiscoveryLoop sftEnv
+  -- We wait for at least two lookups to be sure that the lookup loop looped at
+  -- least once
+  void $ retryEvery10MicrosWhileN 2000 (<= 1) (length <$> readIORef (fakeLookupCalls fakeDNSEnv))
+  Async.cancel discoveryLoop
+
+  actualServers <- readIORef (sftServers sftEnv)
+  assertEqual "servers should be the ones read from DNS" NotDiscoveredYet actualServers
+
+testDiscoveryLoopWhenUnsuccessfulAfterSuccess :: IO ()
+testDiscoveryLoopWhenUnsuccessfulAfterSuccess = do
+  sftEnv <- testDiscoveryLoopWhenSuccessful
+  previousEntries <- readIORef (sftServers sftEnv)
+
+  -- In the following lines we re-use the 'sftEnv' from a successful lookup to
+  -- replicate what will happen when a dns lookup fails after success
+  failingFakeDNSEnv <- newFakeDNSEnv (\_ -> SrvNotAvailable)
+  discoveryLoop <- Async.async $ runM . ignoreLogs . runFakeDNSLookup failingFakeDNSEnv $ sftDiscoveryLoop sftEnv
+  -- We wait for at least two lookups to be sure that the lookup loop looped at
+  -- least once
+  void $ retryEvery10MicrosWhileN 2000 (<= 1) (length <$> readIORef (fakeLookupCalls failingFakeDNSEnv))
+  Async.cancel discoveryLoop
+
+  actualServers <- readIORef (sftServers sftEnv)
+  assertEqual "servers shouldn't get overwriten" previousEntries actualServers
+
+testDiscoveryLoopWhenURLsChange :: IO ()
+testDiscoveryLoopWhenURLsChange = do
+  sftEnv <- testDiscoveryLoopWhenSuccessful
+
+  -- In the following lines we re-use the 'sftEnv' from a successful lookup to
+  -- replicate what will happen when a dns lookup returns new URLs
+  let entry1 = SrvEntry 0 0 (SrvTarget "sft4.foo.example.com." 443)
+      entry2 = SrvEntry 0 0 (SrvTarget "sft5.foo.example.com." 443)
+      newEntries = (entry1 :| [entry2])
+
+  fakeDNSEnv <- newFakeDNSEnv (\_ -> SrvAvailable newEntries)
+  discoveryLoop <- Async.async $ runM . ignoreLogs . runFakeDNSLookup fakeDNSEnv $ sftDiscoveryLoop sftEnv
+  void $ retryEvery10MicrosWhileN 2000 (== 0) (length <$> readIORef (fakeLookupCalls fakeDNSEnv))
+  -- We don't want to stop the loop before it has written to the sftServers IORef
+  void $ retryEvery10MicrosWhileN 2000 (== Discovered newEntries) (readIORef (sftServers sftEnv))
+  Async.cancel discoveryLoop
```

Comment:
Make this a helper function?  `runDiscoveryLoopOnce`?  That would (a) make the tests more concise and (b) also explain a lot better what's going on than just the actual code.

Reply:
it was just pointed out to me that this is slightly trickier than i thought: not only env changes between calls, but also conditions of success.  (also "once" is not a good name, since it loops *until* certain conditions are met.)